core/src/main/java/com/linecorp/armeria/server/Route.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.MediaType;
+
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.Tag;
+
+/**
+ * Matches the {@link RoutingContext} and extracts the path parameters from it.
+ */
+public interface Route {
+
+    /**
+     * Returns a new builder.
+     */
+    static RouteBuilder builder() {
+        return new RouteBuilder();
+    }
+
+    /**
+     * Matches the specified {@link RoutingContext} and extracts the path parameters from it if exists.
+     *
+     * @param routingCtx a context to find the {@link Service}
+     *
+     * @return a non-empty {@link RoutingResult} if the {@linkplain RoutingContext#path() path},
+     *         {@linkplain RoutingContext#method() method},
+     *         {@linkplain RoutingContext#contentType() contentType} and
+     *         {@linkplain RoutingContext#acceptTypes() acceptTypes} matches the equivalent conditions in
+     *         {@link Route}. {@link RoutingResult#empty()} otherwise.
+     *
+     * @see RouteBuilder#methods(Iterable)
+     * @see RouteBuilder#consumes(Iterable)
+     * @see RouteBuilder#produces(Iterable)
+     */
+    RoutingResult apply(RoutingContext routingCtx);
+
+    /**
+     * Returns the names of the path parameters extracted by this mapping.
+     */
+    Set<String> paramNames();
+
+    /**
+     * Returns the logger name.
+     *
+     * @return the logger name whose components are separated by a dot (.)
+     */
+    String loggerName();
+
+    /**
+     * Returns the value of the {@link Tag} in a {@link Meter} of this {@link Route}.
+     */
+    String meterTag();
+
+    /**
+     * Returns the exact path of this {@link Route} if this has an exact path matching condition,
+     * {@link Optional#empty()} otherwise.
+     */
+    Optional<String> exactPath();
+
+    /**
+     * Returns the prefix of this {@link Route} if this has a prefix matching condition,
+     * {@link Optional#empty()} otherwise.
+     *
+     * @return the prefix that ends with '/' if this {@link Route} has a prefix matching condition
+     */
+    Optional<String> prefix();
+
+    /**
+     * Returns the trie path of this {@link Route} if this has a trie matching condition,
+     * {@link Optional#empty()} otherwise. The trie path does not exist when the {@link Route} is created
+     * using a glob or a regex expression.
+     *
+     * @see RouteBuilder#glob(String)
+     * @see RouteBuilder#regex(String)
+     */
+    Optional<String> triePath();
```

Comment:
A lot of these methods look pretty complicated - is it appropriate for them to be in the user-facing API?

Reply:
They are needed in DocService which is in the different package. Let me see what I can do. 🤔

Reply:
The methods are used in various other packages, so can't just hide them. I'm thinking to add a new internal interface or move the `PathMapping` to internal package so we can just expose it. But it takes a little time and will make this  PR bigger so let me do that following PR. 😉

Reply:
Cool - thanks for thinking about it.